### ISSUE TITLE
Creating a tiny C task library

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8253]
Creating a tiny C++ task library<br>
<br>
https://ift.tt/3HZ3YlM<br>
<br>
Vladimir M<br>
<br>
<div>
<div>
<div>
<div>
<div>
<p>Tasks are a popular concurrent programming mechanism. It allows decomposing a computational problem into a set of isolated sub-problems, i.e. tasks. This decomposition exposes the underlying parallelism of the problem. Meaning that at least some tasks can be executed in parallel, and without any explicit synchronization.</p>
<p>This post covers the design and implementation of a very simple yet useful task library.</p>
<p>Tasks are used both for parallelism (CPU-intensive tasks, think parallel sorting) and concurrency (latency-reducing tasks, think responsive UI).</p>
<p>The beauty of tasks is their composability. One task might compute a value, while others depend on that value. Generally speaking, tasks make up a directed acyclic graph, or DAG.</p>
<p>A task library must consist of two parts. An API to specify the task DAG, and a scheduler that actually executes that DAG.</p>
<p>I will start by describing the desired features of the task library and then explain their implementation.</p>
<h1>Creating tasks</h1>
<p>There are several essential operations that pop up in almost any task-based program.</p>
<p>Creating a new task is an obvious first step. So we need a task type to work with.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>class</span> <span>Task</span><span>;</span>
</code></pre>
</div>
</div>
<p>A task is parameterized by the type of the value it computes. A task computing a factorial might have the type <code>Task&lt;int&gt;</code>, and a task requesting a file path from the user&ndash;<code>Task&lt;std::string&gt;</code>.</p>
<p>It is often possible for a task to outlive the scope where it has been constructed. This means that pointers and dynamic allocation are used in most cases. One way to simplify the memory management would be implementing <code>Task</code> in a way similar to the pimpl idiom, making its state shared among (cheap) copies. Such a feature might be useful and desirable, but it complicates the implementation. For now, I simply define a helper pointer alias to save some typing.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>using</span> <span>SPTask</span> <span>=</span> <span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>Task</span><span>&lt;</span><span>T</span><span>&gt;&gt;</span><span>;</span>
</code></pre>
</div>
</div>
<p>To create a new task instance, the <code>make_task</code> function is used.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>F</span><span>,</span> <span>class</span><span>...</span> <span>Ts</span><span>&gt;</span>
<span>SPTask</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>make_task</span><span>(</span><span>F</span> <span>function</span><span>,</span> <span>Ts</span><span>...</span> <span>tasks</span><span>);</span>
</code></pre>
</div>
</div>
<p>Using a free function instead of constructor makes it possible to create an instance of a class derived from <code>Task&lt;T&gt;</code>. This gives more freedom to the implementation.</p>
<p><code>make_task</code> takes as an argument a function that does the computation, and a list of predecessor tasks&ndash;dependencies, which results will be passed on to <code>function</code>.</p>
<p>For example:</p>
<div>
<div>
<pre><code><span>auto</span> <span>getInt</span> <span>=</span> <span>[</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>1</span><span>;</span> <span>};</span>
<span>auto</span> <span>task</span> <span>=</span> <span>make_task</span><span>(</span><span>getInt</span><span>);</span>
</code></pre>
</div>
</div>
<p>This is the simplest use of <code>make_task</code>. <code>getInt</code> doesn&rsquo;t take any parameters, and as such the resulting task does not depend on any other task.</p>
<p>Here is a more complex example:</p>
<div>
<div>
<pre><code><span>auto</span> <span>get3</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>3</span><span>;</span> <span>});</span>
<span>auto</span> <span>get5</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>5</span><span>;</span> <span>});</span>
<span>auto</span> <span>sum</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>(</span><span>int</span> <span>i</span><span>,</span> <span>int</span> <span>j</span><span>)</span> <span>{</span> <span>return</span> <span>i</span> <span>+</span> <span>j</span><span>;</span> <span>},</span> <span>get3</span><span>,</span> <span>get5</span><span>);</span>
<span>auto</span> <span>result</span> <span>=</span> <span>sum</span><span>-&gt;</span><span>result</span><span>();</span> <span>// result == 8</span>
</code></pre>
</div>
</div>
<p>Now there is a dependency. The <code>sum</code> task will only start its execution once both <code>get3</code> and <code>get5</code> complete. But <code>get3</code> and <code>get5</code> do not depend on each other and can execute in parallel.</p>
<p>The last line in the snippet blocks until all the tasks complete and gets the result from <code>sum</code>.</p>
<p>One important case is creating a task inside a task. This might be conditional on the result of a previously completed task.</p>
<div>
<div>
<pre><code><span>auto</span> <span>task</span> <span>=</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>1</span><span>;</span> <span>};</span> <span>});</span>
</code></pre>
</div>
</div>
<p>What type should <code>task</code> have? <code>SPTask&lt;SPTask&lt;int&gt;&gt;</code> doesn&rsquo;t look very useful. It would be quite a chore to compose it with other tasks. For now let&rsquo;s assume that <code>make_task</code> is smart enough to collapse <code>SPTask&lt;SPTask&lt;int&gt;&gt;</code> into <code>SPTask&lt;int&gt;</code> automatically:</p>
<div>
<div>
<pre><code><span>SPTask</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>task</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>1</span><span>;</span> <span>};</span> <span>});</span>
<span>int</span> <span>result</span> <span>=</span> <span>task</span><span>-&gt;</span><span>result</span><span>();</span> <span>// result == 1</span>
</code></pre>
</div>
</div>
<p>Another useful feature is the ability to specify dependencies at runtime in addition to compile time. For example:</p>
<div>
<div>
<pre><code><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>int</span><span>&gt;&gt;</span> <span>tasks</span><span>;</span> <span>// already created</span>
<span>SPTask</span><span>&lt;</span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>int</span><span>&gt;&gt;</span> <span>combined</span> <span>=</span> <span>when_all</span><span>(</span><span>tasks</span><span>);</span>
<span>auto</span> <span>sum</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>(</span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>ints</span><span>)</span> <span>{</span>
    <span>return</span> <span>std</span><span>::</span><span>accumulate</span><span>(</span><span>begin</span><span>(</span><span>ints</span><span>),</span> <span>end</span><span>(</span><span>ints</span><span>),</span> <span>0</span><span>);</span>
<span>},</span> <span>combined</span><span>);</span>
</code></pre>
</div>
</div>
<p>Note the conversion from <code>std::vector&lt;SPTask&lt;int&gt;&gt;</code> to <code>SPTask&lt;std::vector&lt;int&gt;&gt;</code>. This operation is very common and I feel that providing it is essential even for this simple task library.</p>
<p><code>make_task</code> and <code>when_all</code> is our complete task-creating API.</p>
<h1>Implementing the scheduler</h1>
<p>The scheduler does not need to know about task results and their types. It is also oblivious to the dependencies among tasks. This allows for a trivial thread pool implementation. A simpler task interface than <code>Task&lt;T&gt;</code> is sufficient.</p>
<div>
<div>
<pre><code><span>struct</span> <span>ITask</span> <span>{</span>
    <span>virtual</span> <span>void</span> <span>operator</span><span>()()</span> <span>=</span> <span>0</span><span>;</span>
    <span>virtual</span> <span>~</span><span>ITask</span><span>()</span> <span>=</span> <span>default</span><span>;</span>
<span>};</span>
</code></pre>
</div>
</div>
<p>The scheduler has a queue to which new tasks are continuously pushed. Also, there is a predetermined number of threads, that extract tasks from the queue and execute them one by one.</p>
<p><img alt="Task Scheduler" title="Task Scheduler" src="https://rcebits.com/assets/task-scheduler.svg"></p>
<p>The task DAG is assumed to be fed into the queue in topological order (dependencies first). How this can be achieved is the most interesting part of the library and I will discuss it in detail later.</p>
<div>
<div>
<pre><code><span>class</span> <span>Scheduler</span> <span>{</span>
    <span>static</span> <span>std</span><span>::</span><span>unique_ptr</span><span>&lt;</span><span>Scheduler</span><span>&gt;</span> <span>_instance</span><span>;</span>
    <span>Queue</span><span>&lt;</span><span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>ITask</span><span>&gt;&gt;</span> <span>_queue</span><span>;</span>
    <span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>std</span><span>::</span><span>thread</span><span>&gt;</span> <span>_threads</span><span>;</span>

    <span>Scheduler</span><span>();</span>

<span>public:</span>
    <span>static</span> <span>Scheduler</span><span>*</span> <span>instance</span><span>();</span>
    <span>static</span> <span>void</span> <span>init</span><span>();</span>
    <span>static</span> <span>void</span> <span>shutdown</span><span>();</span>
    <span>void</span> <span>schedule</span><span>(</span><span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>ITask</span><span>&gt;</span> <span>task</span><span>);</span>
<span>};</span>
</code></pre>
</div>
</div>
<p>In addition to <code>_queue</code> and <code>_threads</code>, there is a static <code>_instance</code>. <code>init</code> and <code>shutdown</code> need to be called at the startup and shutdown of the application, respectively.</p>
<p>The constructor initializes a thread pool</p>
<div>
<div>
<pre><code><span>Scheduler</span><span>::</span><span>Scheduler</span><span>()</span> <span>{</span>
    <span>auto</span> <span>body</span> <span>=</span> <span>[</span><span>&amp;</span><span>]</span> <span>{</span>
        <span>for</span> <span>(;;)</span> <span>{</span>
            <span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>ITask</span><span>&gt;</span> <span>task</span><span>;</span>
            <span>_queue</span><span>.</span><span>pop</span><span>(</span><span>task</span><span>);</span>
            <span>if</span> <span>(</span><span>!</span><span>task</span><span>)</span>
                <span>return</span><span>;</span>
            <span>(</span><span>*</span><span>task</span><span>)();</span>
        <span>}</span>
    <span>};</span>

    <span>_threads</span><span>.</span><span>resize</span><span>(</span><span>std</span><span>::</span><span>thread</span><span>::</span><span>hardware_concurrency</span><span>());</span>
    <span>for</span> <span>(</span><span>auto</span><span>&amp;</span> <span>th</span> <span>:</span> <span>_threads</span><span>)</span> <span>{</span>
        <span>th</span> <span>=</span> <span>std</span><span>::</span><span>thread</span><span>(</span><span>body</span><span>);</span>
    <span>}</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Each thread continues to execute tasks until it receives a <code>nullptr</code>.</p>
<p>The single instance is created by <code>init</code></p>
<div>
<div>
<pre><code><span>void</span> <span>Scheduler</span><span>::</span><span>init</span><span>()</span> <span>{</span>
    <span>_instance</span><span>.</span><span>reset</span><span>(</span><span>new</span> <span>Scheduler</span><span>());</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>And the threads are finally joined by <code>shutdown</code></p>
<div>
<div>
<pre><code><span>void</span> <span>Scheduler</span><span>::</span><span>shutdown</span><span>()</span> <span>{</span>
    <span>auto</span> <span>i</span> <span>=</span> <span>_instance</span><span>.</span><span>get</span><span>();</span>

    <span>for</span> <span>(</span><span>auto</span> <span>j</span> <span>=</span> <span>0u</span><span>;</span> <span>j</span> <span>&lt;</span> <span>i</span><span>-&gt;</span><span>_threads</span><span>.</span><span>size</span><span>();</span> <span>++</span><span>j</span><span>)</span> <span>{</span>
        <span>i</span><span>-&gt;</span><span>_queue</span><span>.</span><span>push</span><span>({});</span>
    <span>}</span>

    <span>for</span> <span>(</span><span>auto</span><span>&amp;</span> <span>th</span> <span>:</span> <span>i</span><span>-&gt;</span><span>_threads</span><span>)</span> <span>{</span>
        <span>th</span><span>.</span><span>join</span><span>();</span>
    <span>}</span>
<span>}</span>
</code></pre>
</div>
</div>
<p><code>schedule</code> simply pushes the provided task onto the queue</p>
<div>
<div>
<pre><code><span>void</span> <span>Scheduler</span><span>::</span><span>schedule</span><span>(</span><span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>ITask</span><span>&gt;</span> <span>task</span><span>)</span> <span>{</span>
    <span>_queue</span><span>.</span><span>push</span><span>(</span><span>task</span><span>);</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>The <code>Queue</code> type is a simple (and inefficient) thread-safe wrapper of <code>std::queue</code></p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>class</span> <span>Queue</span> <span>{</span>
    <span>std</span><span>::</span><span>queue</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>_queue</span><span>;</span>
    <span>std</span><span>::</span><span>mutex</span> <span>_m</span><span>;</span>
    <span>std</span><span>::</span><span>condition_variable</span> <span>_cv</span><span>;</span>

<span>public:</span>
    <span>void</span> <span>push</span><span>(</span><span>T</span> <span>value</span><span>)</span> <span>{</span>
        <span>auto</span> <span>lock</span> <span>=</span> <span>std</span><span>::</span><span>unique_lock</span><span>(</span><span>_m</span><span>);</span>
        <span>_queue</span><span>.</span><span>push</span><span>(</span><span>value</span><span>);</span>
        <span>_cv</span><span>.</span><span>notify_one</span><span>();</span>
    <span>}</span>

    <span>void</span> <span>pop</span><span>(</span><span>T</span><span>&amp;</span> <span>value</span><span>)</span> <span>{</span>
        <span>auto</span> <span>lock</span> <span>=</span> <span>std</span><span>::</span><span>unique_lock</span><span>(</span><span>_m</span><span>);</span>
        <span>_cv</span><span>.</span><span>wait</span><span>(</span><span>lock</span><span>,</span> <span>[</span><span>this</span><span>]</span> <span>{</span> <span>return</span> <span>!</span><span>_queue</span><span>.</span><span>empty</span><span>();</span> <span>});</span>
        <span>value</span> <span>=</span> <span>_queue</span><span>.</span><span>front</span><span>();</span>
        <span>_queue</span><span>.</span><span>pop</span><span>();</span>
    <span>}</span>
<span>};</span>
</code></pre>
</div>
</div>
<p>This is the scheduler in its entirety.</p>
<p>One question remains. If Scheduler doesn&rsquo;t know anything about task dependencies and blindly schedules tasks on the thread pool, then who does the actual blocking? The answer is: tasks. Tasks themselves are responsible for synchronization.</p>
<h1>Scheduling strategies</h1>
<p>If we had a separate thread for each task, then we would not need to worry about scheduling at all. No matter how many tasks are waiting for their dependencies, they could be certain that sooner or later those dependencies would complete.</p>
<p>Unfortunately, allocating a separate thread for every task can easily become prohibitively expensive, especially on the platforms where creating new threads is costly. Reusing existing threads is a better idea.</p>
<p>As discussed previously, a thread pool implementation requires the task DAG to be processed in topological order. Otherwise deadlocks might occur. To guarantee the topological order, tasks need to be scheduled with care.</p>
<p>Next I will discuss the simple immediate scheduling strategy and show why it fails to guarantee the right scheduling order. Then I will explore a more complex strategy, that only schedules tasks when all their dependencies have completed.</p>
<p>In this strategy tasks go through a series of states:</p>
<ul>
<li>creation and immediate scheduling (pushing onto the thread pool queue) by <code>make_task</code>
</li>
<li>blocking while waiting for dependencies and then executing (both inside <code>operator()</code>)</li>
<li>being in a completed state, holding the result until the last shared pointer has been destroyed</li>
</ul>
<p>This implementation is very simple, but it creates two serious problems.</p>
<p>First, the waiting for the dependencies is done on the thread pool, meaning that while waiting, no useful work can be done by the occupied thread. The thread executing such a task simply yields to the OS.</p>
<p>Second, and more important, is a potential for deadlocks when tasks are created inside other tasks. To understand how this is possible consider the following code:</p>
<div>
<div>
<pre><code><span>auto</span> <span>getIntTask1</span> <span>=</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>0</span><span>;</span> <span>});</span>
<span>auto</span> <span>getIntTask2</span> <span>=</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>3</span><span>;</span> <span>});</span>
<span>auto</span> <span>task</span> <span>=</span> <span>make_task</span><span>([](</span><span>int</span> <span>x</span><span>,</span> <span>int</span> <span>y</span><span>)</span> <span>{</span>
    <span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>Task</span><span>&lt;</span><span>int</span><span>&gt;&gt;&gt;</span> <span>tasks</span><span>;</span>
    <span>for</span> <span>(</span><span>int</span> <span>i</span> <span>=</span> <span>x</span><span>;</span> <span>i</span> <span>&lt;</span> <span>y</span><span>;</span> <span>++</span><span>i</span><span>)</span> <span>{</span>
        <span>tasks</span><span>.</span><span>push_back</span><span>(</span><span>make_task</span><span>([</span><span>=</span><span>]{</span> <span>return</span> <span>i</span> <span>*</span> <span>2</span><span>;</span> <span>}));</span>
    <span>}</span>
    <span>return</span> <span>when_all</span><span>(</span><span>tasks</span><span>);</span>
<span>},</span> <span>getIntTask1</span><span>,</span> <span>getIntTask2</span><span>);</span>

<span>auto</span> <span>sumTask</span> <span>=</span> <span>make_task</span><span>([](</span><span>auto</span> <span>vec</span><span>)</span> <span>{</span>
    <span>return</span> <span>std</span><span>::</span><span>accumulate</span><span>(</span><span>begin</span><span>(</span><span>vec</span><span>),</span> <span>end</span><span>(</span><span>vec</span><span>),</span> <span>0</span><span>);</span>
<span>},</span> <span>task</span><span>);</span>

<span>assert</span><span>(</span><span>sumTask</span><span>-&gt;</span><span>result</span><span>()</span> <span>==</span> <span>6</span><span>);</span>
</code></pre>
</div>
</div>
<p><code>task</code> waits for <code>getIntTask1</code> and <code>getIntTask2</code> to complete, then, based on the result, creates three new tasks and waits for their completion, producing a vector of squares.</p>
<p><code>sumTask</code> waits for <code>task</code> to complete and sums up the resulting squares.</p>
<p>If there are at least two threads in the thread pool, then everything is fine. But if there is only one thread, a deadlock is possible. Furthermore, it is subject to timing.</p>
<p>Look at the Queue (a) below. Arrows show the dependencies among tasks. The dashed rectangles are tasks created dynamically by <code>task</code>.</p>
<p><img alt="Dynamic Task Scheduling" title="Dynamic Task Scheduling" src="https://rcebits.com/assets/task-dynamic.svg"></p>
<p>Both <code>getIntTask1</code> and <code>getIntTask2</code> are scheduled immediately. Then <code>task</code> is scheduled. So far so good, <code>task</code> has not yet created new tasks, including the new task created inside <code>when_all</code>. <code>sumTask</code> gets scheduled next, after <code>task</code>, but before the <code>when_all</code> task.</p>
<p>This leads to a situation in that <code>getIntTask1</code>, <code>getIntTask2</code> and <code>task</code> complete, and then <code>sumTask</code>&rsquo;s <code>operator()</code> is called by the scheduler and <code>sumTask</code> blocks, waiting for tasks 1, 2, 3 and <code>when_all</code>. But none of them are ever going to be executed, because <code>task</code> now occupies the only thread in the thread pool. A deadlock.</p>
<p>If, on the other hand, <code>task</code> completes before <code>sumTask</code> has been scheduled, then all is well. <code>sumTask</code> would be placed after <code>when_all</code> and no deadlock occurs.</p>
<p>This situation is particularly insidious given that the number of threads in the thread pool might not be known upfront.</p>
<p>Sometimes it is possible to carefully rewrite the code to guarantee the correct schedule ordering</p>
<div>
<div>
<pre><code><span>auto</span> <span>getIntTask1</span> <span>=</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>0</span><span>;</span> <span>});</span>
<span>auto</span> <span>getIntTask2</span> <span>=</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>3</span><span>;</span> <span>});</span>
<span>auto</span> <span>task</span> <span>=</span> <span>make_task</span><span>([]</span> <span>(</span><span>int</span> <span>x</span><span>,</span> <span>int</span> <span>y</span><span>)</span> <span>{</span>
    <span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>int</span><span>&gt;&gt;</span> <span>tasks</span><span>;</span>
    <span>for</span> <span>(</span><span>int</span> <span>i</span> <span>=</span> <span>x</span><span>;</span> <span>i</span> <span>&lt;</span> <span>y</span><span>;</span> <span>++</span><span>i</span><span>)</span> <span>{</span>
        <span>tasks</span><span>.</span><span>push_back</span><span>(</span><span>make_task</span><span>([</span><span>=</span><span>]{</span> <span>return</span> <span>i</span> <span>*</span> <span>2</span><span>;</span> <span>}));</span>
    <span>}</span>
    <span>return</span> <span>make_task</span><span>([]</span> <span>(</span><span>auto</span> <span>vec</span><span>)</span> <span>{</span>
        <span>return</span> <span>std</span><span>::</span><span>accumulate</span><span>(</span><span>begin</span><span>(</span><span>vec</span><span>),</span> <span>end</span><span>(</span><span>vec</span><span>),</span> <span>0</span><span>);</span>
    <span>},</span> <span>when_all</span><span>(</span><span>tasks</span><span>));</span>
<span>},</span> <span>getIntTask1</span><span>,</span> <span>getIntTask2</span><span>);</span>

<span>assert</span><span>(</span><span>task</span><span>-&gt;</span><span>result</span><span>()</span> <span>==</span> <span>6</span><span>);</span>
</code></pre>
</div>
</div>
<p>This code will always create the ordering shown in Queue (b) above.</p>
<p>There are ways to automatically resolve such conflicts. For example, in addition to the three states <em>scheduled, executing</em> and <em>completed</em>, a new <em>unscheduled</em> state can be introduced. A task with dependencies would not be placed into the queue immediately, but instead each of its dependencies would carry a pointer to it. And the last dependency to complete would trigger the scheduling of the unscheduled task. Such design is safer, but requires additional synchronization and bookkeeping.</p>
<p>Any serious task library should offer a solution to this problem.</p>
<h1>Scheduling strategy 2. Continuation-style scheduling.</h1>
<p>In this strategy a task is scheduled after the last of its dependencies has been scheduled or executed. If a task does not have any dependencies, it is scheduled immediately upon creation. This solves the problem discussed in the previous section. It makes code independent of the thread pool size and removes the possibility of deadlocks. The downside of this strategy is added complexity.</p>
<p>Any task with dependencies becomes a continuation. Once a continuation is created, it needs to be attached to all its dependencies.</p>
<p>Consider the following code</p>
<div>
<div>
<pre><code><span>auto</span> <span>task1</span> <span>=</span> <span>make_task</span><span>([]</span> <span>{</span>
    <span>// long computation</span>
    <span>int</span> <span>a</span> <span>=</span> <span>1</span><span>;</span>
    <span>return</span> <span>make_task</span><span>([</span><span>a</span><span>]</span> <span>{</span>     <span>// task2</span>
        <span>// long computation</span>
        <span>int</span> <span>b</span> <span>=</span> <span>a</span> <span>+</span> <span>1</span><span>;</span>
        <span>return</span> <span>make_task</span><span>([</span><span>b</span><span>]</span> <span>{</span> <span>// task3</span>
            <span>return</span> <span>b</span> <span>+</span> <span>1</span><span>;</span>
        <span>});</span>
    <span>});</span>
<span>});</span>

<span>auto</span> <span>task4</span> <span>=</span> <span>make_task</span><span>([]</span> <span>(</span><span>int</span> <span>i</span><span>)</span> <span>{</span>
    <span>return</span> <span>i</span> <span>*</span> <span>i</span><span>;</span>
<span>},</span> <span>task1</span><span>);</span>

<span>// task4-&gt;result() == 9</span>
</code></pre>
</div>
</div>
<p>Visually the task DAG looks like this</p>
<p><img alt="Dynamic Tasks" title="Dynamic Tasks" src="https://rcebits.com/assets/task-scheduler-deps.svg"></p>
<p>(Dashed outlines mean that the tasks are created by other tasks.)</p>
<p>There are two essential properties that we want to maintain:</p>
<ol>
<li>
<code>task4</code> must await <code>task3</code> and not <code>task1</code>.</li>
<li>
<code>task4</code> should only be scheduled after <code>task3</code>.</li>
</ol>
<p>The first property is guaranteed simply by blocking until the dependency results are available. This property does not depend on scheduling.</p>
<p>The second property is harder to achieve.</p>
<p>In the example above it is assumed that <code>task1</code> (without considering the tasks it creates) does a long computation before it finally creates and returns <code>task2</code>. This means that <code>task4</code> is almost certainly created before <code>task2</code> and <code>task3</code>. At this point <code>task1</code> is either waiting in the scheduler queue (<em>scheduled</em>) or is being executed (<em>running</em>). In both cases we do not want to schedule <code>task4</code> immediately. We want to wait until <code>task3</code> has been scheduled.</p>
<p>After having created <code>task4</code>, we attach the new task to <code>task1</code> and do not schedule it. Once <code>task1</code> completes, it still does not schedule <code>task4</code>, instead, it reattaches <code>task4</code> to <code>task2</code>. When <code>task2</code> completes, it passes <code>task4</code> further down to <code>task3</code>. And only when <code>task3</code> completes, it finally schedules <code>task4</code> (because <code>task3</code> does not return any new tasks).</p>
<p>This is one possible sequence of events. It is also possible that by the time <code>task4</code> is created, the whole chain <code>task1</code> -&gt; <code>task2</code> -&gt; <code>task3</code> has already completed, which means that <code>task4</code> must be scheduled immediately upon creation. Another possibility is that <code>task4</code> is created after <code>task1</code> has completed, but <code>task2</code> has not. In this case <code>task4</code> must be attached to <code>task2</code>. In general, it means that to which task a dependency gets attached is decided at runtime and is subject to timing.</p>
<p>To implement this strategy, a new class is created, an instance of which is held by all tasks</p>
<div>
<div>
<pre><code><span>class</span> <span>ContinuationHolder</span> <span>{</span>
    <span>std</span><span>::</span><span>mutex</span> <span>_m</span><span>;</span>
    <span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>std</span><span>::</span><span>function</span><span>&lt;</span><span>void</span><span>()</span><span>&gt;&gt;</span> <span>_actions</span><span>;</span>
    <span>bool</span> <span>_scheduled</span> <span>=</span> <span>false</span><span>;</span>
    <span>ContinuationHolder</span><span>*</span> <span>_transferred</span> <span>=</span> <span>nullptr</span><span>;</span>

<span>public:</span>
    <span>void</span> <span>attach</span><span>(</span><span>std</span><span>::</span><span>function</span><span>&lt;</span><span>void</span><span>()</span><span>&gt;</span> <span>action</span><span>);</span>
    <span>void</span> <span>schedule</span><span>();</span>
    <span>void</span> <span>transfer</span><span>(</span><span>ContinuationHolder</span><span>*</span> <span>dest</span><span>);</span>
<span>};</span>
</code></pre>
</div>
</div>
<p>Instead of attaching actual <code>ITask</code>s, a <code>std::function</code> is used to reduce coupling.</p>
<p><code>ContinuationHolder</code> has three states, that are related to the state of the associated task:</p>
<ol>
<li>Accepting new continuations (the task is <em>scheduled</em> or <em>running</em>). New actions in this state are being added to a list.</li>
<li>Scheduled (the task has completed and not returned a new task). New actions are executed immediately.</li>
<li>Transferred (the task has completed and returned a new task, to which it reattached its continuations). New actions are delegated to the next instance of <code>ContinuationHolder</code>.</li>
</ol>
<p>The implementation is straightforward. <code>schedule</code> changes the state and executes all previously attached actions:</p>
<div>
<div>
<pre><code><span>void</span> <span>ContinuationHolder</span><span>::</span><span>schedule</span><span>()</span> <span>{</span>
    <span>{</span>
        <span>auto</span> <span>lock</span> <span>=</span> <span>std</span><span>::</span><span>lock_guard</span><span>(</span><span>_m</span><span>);</span>
        <span>assert</span><span>(</span><span>!</span><span>_scheduled</span><span>);</span>
        <span>_scheduled</span> <span>=</span> <span>true</span><span>;</span>
    <span>}</span>
    <span>for</span> <span>(</span><span>auto</span><span>&amp;</span> <span>action</span> <span>:</span> <span>_actions</span><span>)</span> <span>{</span>
        <span>action</span><span>();</span>
    <span>}</span>
<span>}</span>
</code></pre>
</div>
</div>
<p><code>transfer</code> works similarly</p>
<div>
<div>
<pre><code><span>void</span> <span>ContinuationHolder</span><span>::</span><span>transfer</span><span>(</span><span>ContinuationHolder</span><span>*</span> <span>dest</span><span>)</span> <span>{</span>
    <span>{</span>
        <span>auto</span> <span>lock</span> <span>=</span> <span>std</span><span>::</span><span>lock_guard</span><span>(</span><span>_m</span><span>);</span>
        <span>assert</span><span>(</span><span>!</span><span>_scheduled</span><span>);</span>
        <span>_transferred</span> <span>=</span> <span>dest</span><span>;</span>
    <span>}</span>
    <span>for</span> <span>(</span><span>auto</span><span>&amp;</span> <span>action</span> <span>:</span> <span>_actions</span><span>)</span> <span>{</span>
        <span>dest</span><span>-&gt;</span><span>attach</span><span>(</span><span>action</span><span>);</span>
    <span>}</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>And <code>attach</code> dispatches <code>action</code> based on the state:</p>
<div>
<div>
<pre><code><span>void</span> <span>ContinuationHolder</span><span>::</span><span>attach</span><span>(</span><span>std</span><span>::</span><span>function</span><span>&lt;</span><span>void</span><span>()</span><span>&gt;</span> <span>action</span><span>)</span> <span>{</span>
    <span>auto</span> <span>lock</span> <span>=</span> <span>std</span><span>::</span><span>lock_guard</span><span>(</span><span>_m</span><span>);</span>
    <span>if</span> <span>(</span><span>_transferred</span><span>)</span> <span>{</span>
        <span>_transferred</span><span>-&gt;</span><span>attach</span><span>(</span><span>action</span><span>);</span>
    <span>}</span> <span>else</span> <span>if</span> <span>(</span><span>_scheduled</span><span>)</span> <span>{</span>
        <span>action</span><span>();</span>
    <span>}</span> <span>else</span> <span>{</span>
        <span>_actions</span><span>.</span><span>push_back</span><span>(</span><span>action</span><span>);</span>
    <span>}</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Now we have a suitable scheduling strategy and the tools to finally implement <code>Task&lt;T&gt;</code>.</p>
<h1>Implementing Task&lt;T&gt;</h1>
<p>The implementation will be split between <code>Task&lt;T&gt;</code> and <code>TaskImpl&lt;F, Ps...&gt;</code>. <code>Task&lt;T&gt;</code> is the user-facing part. It is abstract (it doesn&rsquo;t implement <code>operator()</code>) and never created directly.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>class</span> <span>Task</span> <span>:</span> <span>public</span> <span>ITask</span><span>,</span> <span>public</span> <span>std</span><span>::</span><span>enable_shared_from_this</span><span>&lt;</span><span>Task</span><span>&lt;</span><span>T</span><span>&gt;&gt;</span> <span>{</span>
<span>protected:</span>
    <span>std</span><span>::</span><span>shared_future</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>_future</span><span>;</span>
    <span>std</span><span>::</span><span>atomic</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>_unscheduledDeps</span> <span>=</span> <span>0</span><span>;</span>
    <span>ContinuationHolder</span> <span>_ch</span><span>;</span>

    <span>template</span> <span>&lt;</span><span>class</span> <span>Ff</span><span>,</span> <span>class</span><span>...</span> <span>Psf</span><span>&gt;</span>
    <span>friend</span> <span>class</span> <span>TaskImpl</span><span>;</span>

<span>public:</span>
    <span>using</span> <span>R</span> <span>=</span> <span>T</span><span>;</span>

    <span>Task</span><span>(</span><span>int</span> <span>unscheduledDeps</span><span>)</span> <span>:</span> <span>_unscheduledDeps</span><span>(</span><span>unscheduledDeps</span><span>)</span> <span>{</span> <span>}</span>

    <span>R</span> <span>result</span><span>()</span> <span>{</span>
        <span>return</span> <span>_future</span><span>.</span><span>get</span><span>();</span>
    <span>}</span>

    <span>void</span> <span>wait</span><span>()</span> <span>override</span> <span>{</span>
        <span>_future</span><span>.</span><span>get</span><span>();</span>
    <span>}</span>

    <span>void</span> <span>addOnSchedule</span><span>(</span><span>std</span><span>::</span><span>function</span><span>&lt;</span><span>void</span><span>()</span><span>&gt;</span> <span>action</span><span>)</span> <span>{</span>
        <span>_ch</span><span>.</span><span>attach</span><span>(</span><span>action</span><span>);</span>
    <span>}</span>

    <span>ContinuationHolder</span><span>*</span> <span>ch</span><span>()</span> <span>{</span>
        <span>return</span> <span>&amp;</span><span>_ch</span><span>;</span>
    <span>}</span>

    <span>void</span> <span>signalDependency</span><span>()</span> <span>{</span>
        <span>if</span> <span>(</span><span>!--</span><span>_unscheduledDeps</span><span>)</span> <span>{</span>
            <span>Scheduler</span><span>::</span><span>instance</span><span>()</span><span>-&gt;</span><span>schedule</span><span>(</span><span>this</span><span>-&gt;</span><span>shared_from_this</span><span>());</span>
        <span>}</span>
    <span>}</span>
<span>};</span>
</code></pre>
</div>
</div>
<p>A <em>shared</em> future is used because several tasks might await the same dependency task.</p>
<p><code>_unscheduledDeps</code> is the number of currently unscheduled dependencies. This number is decremented each time a dependency completes. Once the number reaches zero, <code>ContinuationHolder</code> schedules all attached continuations.</p>
<p>The friend declaration is necessary to allow <code>Task&lt;T&gt;</code> access the <code>_future</code> member of another <code>Task&lt;Z&gt;</code>, where <code>T</code> and <code>Z</code> are different types. Say, <code>Task&lt;int&gt;</code> should be able to access <code>Task&lt;float&gt;::_future</code>.</p>
<p><code>Task&lt;T&gt;</code> also exposes the type of its value, <code>R</code>.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>F</span><span>,</span> <span>class</span><span>...</span> <span>Ps</span><span>&gt;</span>
<span>class</span> <span>TaskImpl</span> <span>:</span> <span>public</span> <span>Task</span><span>&lt;</span><span>std</span><span>::</span><span>invoke_result_t</span><span>&lt;</span><span>F</span><span>,</span> <span>typename</span> <span>Ps</span><span>::</span><span>element_type</span><span>::</span><span>R</span><span>...</span><span>&gt;&gt;</span> <span>{</span>
<span>public:</span>
    <span>using</span> <span>R</span> <span>=</span> <span>std</span><span>::</span><span>invoke_result_t</span><span>&lt;</span><span>F</span><span>,</span> <span>typename</span> <span>Ps</span><span>::</span><span>element_type</span><span>::</span><span>R</span><span>...</span><span>&gt;</span><span>;</span>

<span>private:</span>
    <span>std</span><span>::</span><span>promise</span><span>&lt;</span><span>R</span><span>&gt;</span> <span>_promise</span><span>;</span>
    <span>F</span> <span>_body</span><span>;</span>
    <span>std</span><span>::</span><span>tuple</span><span>&lt;</span><span>Ps</span><span>...</span><span>&gt;</span> <span>_deps</span><span>;</span>

<span>public:</span>
    <span>TaskImpl</span><span>(</span><span>int</span> <span>depCount</span><span>,</span> <span>F</span> <span>body</span><span>,</span> <span>Ps</span><span>...</span> <span>ps</span><span>)</span>
        <span>:</span> <span>Task</span><span>&lt;</span><span>R</span><span>&gt;</span><span>(</span><span>depCount</span><span>),</span> <span>_body</span><span>(</span><span>body</span><span>),</span> <span>_deps</span><span>(</span><span>ps</span><span>...)</span> <span>{</span>
        <span>Task</span><span>&lt;</span><span>R</span><span>&gt;::</span><span>_future</span> <span>=</span> <span>_promise</span><span>.</span><span>get_future</span><span>();</span>
    <span>}</span>

    <span>void</span> <span>operator</span><span>()()</span> <span>override</span> <span>{</span>
        <span>auto</span> <span>i</span> <span>=</span> <span>[</span><span>&amp;</span><span>](</span><span>Ps</span><span>...</span> <span>ps</span><span>)</span> <span>{</span>
            <span>_promise</span><span>.</span><span>set_value</span><span>(</span><span>_body</span><span>(</span><span>ps</span><span>-&gt;</span><span>_future</span><span>.</span><span>get</span><span>()...));</span>
        <span>};</span>
        <span>std</span><span>::</span><span>apply</span><span>(</span><span>i</span><span>,</span> <span>_deps</span><span>);</span>
        <span>if</span> <span>constexpr</span> <span>(</span><span>TaskTraits</span><span>&lt;</span><span>R</span><span>&gt;::</span><span>isTask</span><span>)</span> <span>{</span>
            <span>auto</span><span>&amp;</span> <span>task</span> <span>=</span> <span>Task</span><span>&lt;</span><span>R</span><span>&gt;::</span><span>_future</span><span>.</span><span>get</span><span>();</span>
            <span>Task</span><span>&lt;</span><span>R</span><span>&gt;::</span><span>_ch</span><span>.</span><span>transfer</span><span>(</span><span>task</span><span>-&gt;</span><span>ch</span><span>());</span>
        <span>}</span> <span>else</span> <span>{</span>
            <span>Task</span><span>&lt;</span><span>R</span><span>&gt;::</span><span>_ch</span><span>.</span><span>schedule</span><span>();</span>
        <span>}</span>
    <span>}</span>
<span>};</span>
</code></pre>
</div>
</div>
<p>Note that the type <code>T</code> in <code>Task&lt;T&gt;</code> is not specified explicitly in <code>TaskImpl&lt;F, Ps...&gt;</code>. Instead it is the result type of callable <code>F</code>.</p>
<p>The dependencies <code>Ps...</code> are assumed to be shared pointers: <code>std::shared_ptr&lt;Task&lt;T&gt;&gt;</code>. The type <code>T</code> is then computed as</p>
<div>
<div>
<pre><code><span>std</span><span>::</span><span>invoke_result_t</span><span>&lt;</span><span>F</span><span>,</span> <span>typename</span> <span>Ps</span><span>::</span><span>element_type</span><span>::</span><span>R</span><span>...</span><span>&gt;</span>
</code></pre>
</div>
</div>
<p><code>_promise</code> and <code>_body</code> should be self-explanatory.</p>
<p><code>_deps</code> is a heterogeneous container of the task dependencies. Given that the number of dependencies and their types are known at compile time, <code>std::tuple&lt;&gt;</code> is the obvious choice.</p>
<p>The constructor simply initializes all the variables.</p>
<p>The <code>operator()</code> is the heart of <code>TaskImpl&lt;&gt;</code>. It first gets the results of all dependencies, blocking if they are not yet ready. It then passes the results of all dependencies to <code>_body</code>, and stores the returned value inside this task&rsquo;s <code>_promise</code>. Then, depending on the return type of <code>F</code>, either transfers the attached continuations to the result task or schedules them directly.</p>
<p><code>TaskTraits&lt;T&gt;</code> is a simple trait template:</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>struct</span> <span>TaskTraits</span> <span>{</span>
    <span>static</span> <span>constexpr</span> <span>bool</span> <span>isTask</span> <span>=</span> <span>false</span><span>;</span>
<span>};</span>

<span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>struct</span> <span>TaskTraits</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>T</span><span>&gt;&gt;</span> <span>{</span>
    <span>static</span> <span>constexpr</span> <span>bool</span> <span>isTask</span> <span>=</span> <span>true</span><span>;</span>
<span>};</span>
</code></pre>
</div>
</div>
<h1>Implementing make_task</h1>
<p><code>make_task</code> is a utility function that creates an instance of <code>TaskImpl&lt;&gt;</code>, schedules it, and collapses values of type <code>SPTask&lt;SPTask&lt;T&gt;&gt;</code> into <code>SPTask&lt;T&gt;</code>. The last part is done using <code>TaskJoiner</code>:</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>struct</span> <span>TaskJoiner</span> <span>{</span>
    <span>T</span> <span>operator</span><span>()(</span><span>T</span> <span>t</span><span>)</span> <span>{</span>
        <span>return</span> <span>t</span><span>;</span>
    <span>}</span>
<span>};</span>

<span>template</span> <span>&lt;</span><span>class</span> <span>T</span><span>&gt;</span>
<span>struct</span> <span>TaskJoiner</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>T</span><span>&gt;&gt;&gt;</span> <span>{</span>
    <span>auto</span> <span>operator</span><span>()(</span><span>auto</span> <span>task</span><span>)</span> <span>{</span>
        <span>return</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>(</span><span>auto</span> <span>x</span><span>)</span> <span>{</span> <span>return</span> <span>x</span><span>-&gt;</span><span>result</span><span>();</span> <span>},</span> <span>task</span><span>);</span>
    <span>}</span>
<span>};</span>
</code></pre>
</div>
</div>
<p><code>TaskJoiner&lt;&gt;</code> uses partial template specialization to pattern-match <code>SPTask&lt;SPTask&lt;T&gt;&gt;</code>. Notably, in the general case the <code>T</code> type is <code>SPTask&lt;&gt;</code>, while in the specialized case it is not a task at all, but a concrete return type like <code>int</code> or <code>float</code>.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>F</span><span>,</span> <span>class</span><span>...</span> <span>T</span><span>&gt;</span>
<span>auto</span> <span>make_task</span><span>(</span><span>F</span> <span>f</span><span>,</span> <span>T</span><span>...</span> <span>x</span><span>)</span> <span>{</span>
    <span>const</span> <span>auto</span> <span>depCount</span> <span>=</span> <span>sizeof</span><span>...(</span><span>T</span><span>);</span>
    <span>auto</span> <span>task</span> <span>=</span> <span>std</span><span>::</span><span>make_shared</span><span>&lt;</span><span>TaskImpl</span><span>&lt;</span><span>F</span><span>,</span> <span>T</span><span>...</span><span>&gt;&gt;</span><span>(</span><span>depCount</span><span>,</span> <span>f</span><span>,</span> <span>x</span><span>...);</span>
    <span>if</span> <span>constexpr</span> <span>(</span><span>depCount</span> <span>==</span> <span>0</span><span>)</span> <span>{</span>
        <span>Scheduler</span><span>::</span><span>instance</span><span>()</span><span>-&gt;</span><span>schedule</span><span>(</span><span>task</span><span>);</span>
    <span>}</span> <span>else</span> <span>{</span>
        <span>(...,</span> <span>x</span><span>-&gt;</span><span>addOnSchedule</span><span>([</span><span>=</span><span>]</span> <span>{</span>
            <span>task</span><span>-&gt;</span><span>signalDependency</span><span>();</span>
        <span>}));</span>
    <span>}</span>
    <span>using</span> <span>TaskType</span> <span>=</span> <span>SPTask</span><span>&lt;</span><span>typename</span> <span>TaskImpl</span><span>&lt;</span><span>F</span><span>,</span> <span>T</span><span>...</span><span>&gt;::</span><span>R</span><span>&gt;</span><span>;</span>
    <span>return</span> <span>TaskJoiner</span><span>&lt;</span><span>TaskType</span><span>&gt;</span><span>()(</span><span>task</span><span>);</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Now <code>make_task</code> needs some unpacking. First, it always creates a new task, passing to it the number of dependencies. The number of dependencies in this case is known at compile time.</p>
<p>If there are no dependencies, it means the task can be scheduled immediately. Otherwise the newly created task is attached to each dependency in the parameter pack using a folding expression.</p>
<p>The last step applies <code>TaskJoiner&lt;T&gt;</code> to the result type.</p>
<h1>Implementing when_all</h1>
<p><code>when_all</code> is very similar to <code>make_task</code>, except that the dependencies are only known at runtime, and the resulting task is always of type <code>std::vector&lt;T&gt;</code> and as such does not need <code>TaskJoiner&lt;T&gt;</code>.</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>R</span><span>&gt;</span>
<span>auto</span> <span>when_all</span><span>(</span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>R</span><span>&gt;&gt;</span> <span>deps</span><span>)</span> <span>{</span>
    <span>assert</span><span>(</span><span>!</span><span>deps</span><span>.</span><span>empty</span><span>());</span>
    <span>auto</span> <span>collect</span> <span>=</span> <span>[</span><span>=</span><span>]</span> <span>{</span>
        <span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>R</span><span>&gt;</span> <span>vec</span><span>;</span>
        <span>for</span> <span>(</span><span>auto</span> <span>dep</span> <span>:</span> <span>deps</span><span>)</span> <span>{</span>
            <span>vec</span><span>.</span><span>push_back</span><span>(</span><span>dep</span><span>-&gt;</span><span>result</span><span>());</span>
        <span>}</span>
        <span>return</span> <span>vec</span><span>;</span>
    <span>};</span>
    <span>auto</span> <span>task</span> <span>=</span> <span>std</span><span>::</span><span>make_shared</span><span>&lt;</span><span>TaskImpl</span><span>&lt;</span><span>decltype</span><span>(</span><span>collect</span><span>)</span><span>&gt;&gt;</span><span>(</span>
        <span>deps</span><span>.</span><span>size</span><span>(),</span> <span>collect</span><span>);</span>
    <span>for</span> <span>(</span><span>auto</span> <span>dep</span> <span>:</span> <span>deps</span><span>)</span> <span>{</span>
        <span>dep</span><span>-&gt;</span><span>addOnSchedule</span><span>([</span><span>=</span><span>]</span> <span>{</span>
            <span>task</span><span>-&gt;</span><span>signalDependency</span><span>();</span>
        <span>});</span>
    <span>}</span>
    <span>return</span> <span>static_cast</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>std</span><span>::</span><span>vector</span><span>&lt;</span><span>R</span><span>&gt;&gt;&gt;</span><span>(</span><span>task</span><span>);</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Now the complete task-creating API is implemented.</p>
<h1>The lack of exception handling.</h1>
<p>So far I completely ignored exceptions. Any exception inside the thread pool will just terminate the program. One way to handle exceptions is to catch exceptions thrown inside task bodies and put them into <code>std::promise</code>. Then instead of passing unpacked dependency results as arguments, the <code>std::future</code>s could be passed directly. The dependant tasks&rsquo; bodies would then access the passed <code>std::future</code>s and trigger the previously captured exception, allowing them to either handle it, or pass it further down the task DAG.</p>
<p>For the sake of simplicity I omitted exception handling.</p>
<h1>Bonus: It looks suspiciously monadic!</h1>
<p>Indeed, <code>make_task</code> combines the function of monadic <strong>return</strong> and <strong>bind</strong></p>
<div>
<div>
<pre><code><span>bind</span> <span>::</span> <span>M</span> <span>T</span> <span>-&gt;</span> <span>(</span><span>T</span> <span>-&gt;</span> <span>M</span> <span>R</span><span>)</span> <span>-&gt;</span> <span>M</span> <span>R</span>
</code></pre>
</div>
</div>
<p>Which is the same as</p>
<div>
<div>
<pre><code><span>std</span><span>::</span><span>function</span><span>&lt;</span><span>SPTask</span><span>&lt;</span><span>R</span><span>&gt;</span><span>(</span><span>T</span><span>)</span><span>&gt;</span> <span>f</span><span>;</span>
<span>SPTask</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>t</span><span>;</span>
<span>SPTask</span><span>&lt;</span><span>R</span><span>&gt;</span> <span>=</span> <span>make_task</span><span>(</span><span>f</span><span>,</span> <span>t</span><span>);</span>
</code></pre>
</div>
</div>
<p>thanks to the collapsing rule <code>Task&lt;Task&lt;T&gt;&gt;</code> =&gt; <code>Task&lt;T&gt;</code>, which is itself the monadic <strong>join</strong></p>
<p>Of course, there is also the monadic <strong>return</strong></p>
<p>The function of which is achieved by</p>
<div>
<div>
<pre><code><span>SPTask</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>task</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>T</span><span>{};</span> <span>});</span>
</code></pre>
</div>
</div>
<p>I have not checked the monad laws, so I am not sure <code>Task&lt;T&gt;</code> is a true monad, but in any case, it feels like one.</p>
<h1>An example: implementing quicksort using tasks</h1>
<p>Quicksort is an easy algorithm to parallelize. The sequential version looks like this:</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>Iter</span><span>&gt;</span>
<span>Iter</span> <span>qsPartition</span><span>(</span><span>Iter</span> <span>begin</span><span>,</span> <span>Iter</span> <span>end</span><span>)</span> <span>{</span>
    <span>auto</span> <span>i</span> <span>=</span> <span>begin</span><span>,</span> <span>p</span> <span>=</span> <span>end</span> <span>-</span> <span>1</span><span>;</span>
    <span>for</span> <span>(</span><span>auto</span> <span>j</span> <span>=</span> <span>begin</span><span>;</span> <span>j</span> <span>&lt;</span> <span>p</span><span>;</span> <span>++</span><span>j</span><span>)</span> <span>{</span>
        <span>if</span> <span>(</span><span>*</span><span>j</span> <span>&lt;=</span> <span>*</span><span>p</span><span>)</span> <span>{</span>
            <span>std</span><span>::</span><span>iter_swap</span><span>(</span><span>i</span><span>,</span> <span>j</span><span>);</span>
            <span>++</span><span>i</span><span>;</span>
        <span>}</span>
    <span>}</span>
    <span>std</span><span>::</span><span>iter_swap</span><span>(</span><span>p</span><span>,</span> <span>i</span><span>);</span>
    <span>return</span> <span>i</span><span>;</span>
<span>}</span>

<span>template</span> <span>&lt;</span><span>class</span> <span>Iter</span><span>&gt;</span>
<span>void</span> <span>quickSortSeq</span><span>(</span><span>Iter</span> <span>begin</span><span>,</span> <span>Iter</span> <span>end</span><span>)</span> <span>{</span>
    <span>if</span> <span>(</span><span>begin</span> <span>&gt;=</span> <span>end</span><span>)</span>
        <span>return</span><span>;</span>
    <span>auto</span> <span>p</span> <span>=</span> <span>qsPartition</span><span>(</span><span>begin</span><span>,</span> <span>end</span><span>);</span>
    <span>quickSortSeq</span><span>(</span><span>begin</span><span>,</span> <span>p</span><span>),</span>
    <span>quickSortSeq</span><span>(</span><span>p</span> <span>+</span> <span>1</span><span>,</span> <span>end</span><span>);</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>Once the partition step is done, the elements to the left and right of the pivot element can be sorted independently. Here is a parallel version:</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>Iter</span><span>&gt;</span>
<span>SPTask</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>quickSort</span><span>(</span><span>Iter</span> <span>begin</span><span>,</span> <span>Iter</span> <span>end</span><span>)</span> <span>{</span>
    <span>if</span> <span>(</span><span>begin</span> <span>&gt;=</span> <span>end</span><span>)</span>
        <span>return</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>0</span><span>;</span> <span>});</span>
    <span>auto</span> <span>partition</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>qsPartition</span><span>(</span><span>begin</span><span>,</span> <span>end</span><span>);</span> <span>});</span>
    <span>auto</span> <span>recurse</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>(</span><span>auto</span> <span>p</span><span>)</span> <span>{</span>
        <span>return</span> <span>when_all</span><span>(</span><span>std</span><span>::</span><span>vector</span><span>{</span>
            <span>quickSort</span><span>(</span><span>begin</span><span>,</span> <span>p</span><span>),</span>
            <span>quickSort</span><span>(</span><span>p</span> <span>+</span> <span>1</span><span>,</span> <span>end</span><span>)});</span>
    <span>},</span> <span>partition</span><span>);</span>
    <span>return</span> <span>make_task</span><span>([]</span> <span>(</span><span>auto</span><span>)</span> <span>{</span> <span>return</span> <span>0</span><span>;</span> <span>},</span> <span>recurse</span><span>);</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>The partition step is done in a separate task. Once it completes, two new <code>quickSort</code> task are created. When both of them complete, the whole initial invocation of <code>quickSort</code> completes.</p>
<p>If <code>when_all</code> had supported <code>ITask</code> in addition to <code>Task&lt;T&gt;</code> and also accepted variadic packs (and not only <code>std::vector</code>s), we could improve the code, making it more natural:</p>
<div>
<div>
<pre><code><span>template</span> <span>&lt;</span><span>class</span> <span>Iter</span><span>&gt;</span>
<span>std</span><span>::</span><span>shared_ptr</span><span>&lt;</span><span>ITask</span><span>&gt;</span> <span>quickSort</span><span>(</span><span>Iter</span> <span>begin</span><span>,</span> <span>Iter</span> <span>end</span><span>)</span> <span>{</span>
    <span>if</span> <span>(</span><span>begin</span> <span>&gt;=</span> <span>end</span><span>)</span>
        <span>return</span> <span>make_task</span><span>([]</span> <span>{</span> <span>return</span> <span>0</span><span>;</span> <span>});</span>
    <span>auto</span> <span>partition</span> <span>=</span> <span>make_task</span><span>([</span><span>=</span><span>]</span> <span>{</span> <span>return</span> <span>qsPartition</span><span>(</span><span>begin</span><span>,</span> <span>end</span><span>);</span> <span>});</span>
    <span>return</span> <span>make_task</span><span>([]</span> <span>(</span><span>auto</span> <span>p</span><span>)</span> <span>{</span>
        <span>return</span> <span>when_all</span><span>(</span>
            <span>quickSort</span><span>(</span><span>begin</span><span>,</span> <span>p</span><span>),</span>
            <span>quickSort</span><span>(</span><span>p</span> <span>+</span> <span>1</span><span>,</span> <span>end</span><span>));</span>
    <span>},</span> <span>partition</span><span>);</span>
<span>}</span>
</code></pre>
</div>
</div>
<p>As is, this parallel version is unlikely to be faster than the sequential one, due to the overhead created by tasks. A simple way to overcome this problem is to fallback to the sequential version when the array size is small, say 1000 elements or less.</p>
<h1>Conclusion</h1>
<p>In under 300 lines of code, we got ourselves a surprisingly capable task library. There are, of course, many improvements to be made. The queue might benefit from some performance optimization, the API might use some polish, more features such as <code>when_any</code> might be provided, as well as the ability to query the task state: <em>unscheduled</em>, <em>scheduled</em>, <em>running</em>, <em>completed</em>. Making <code>ITask</code> a first-class citizen would allow working more easily with tasks that do not return any value. Exception handling and propagation is also an obvious next step.</p>
<p>Still, the library is done completely using standard C++ and is pretty expressive thanks to the use of some basic template metaprogramming.</p>
<p>The complete code with a few test cases can be found on <a href="https://github.com/nongeneric/tinytasks">github</a>.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via RCE bits<br>
<br>
January 12, 2022 at 07:11PM
